### DISCUSSION TITLE
Request IPs?

### DISCUSSION THREAD

**Discussion by BobbyWibowo** [kartikk221/hyper-express#91] [Q&A] [answered]
Hyper-Express was started with
```js
const safe = new HyperExpress.Server({
  trust_proxy: true
})
```

Yet, here's what I'm getting in Response:
```js
req.headers: {
  connection: 'upgrade',
  host: 'safe.fiery.me',
  'x-real-ip': '180.x.x.x',
  'x-forwarded-for': '180.x.x.x, 162.158.163.156',
  'x-forwarded-proto': 'https',
  ...
}
req.ip: '127.0.0.1'
req.proxy_ip: ''
```
In `x-forwarded-for` is my IP, then followed by Cloudflare IP

It's running behind Nginx and Cloudflare, for context

**Comment by kartikk221** [comment, marked as answer]
Could you clarify how the behavior of the `trust_proxy` parameter is incorrect or unintended?
The `trust_proxy` parameter is internally used in ExpressJS compatibility properties like `Response.protocol`, `Response.ips`, `Response.hostname`.

**Reply by BobbyWibowo** [reply]
I see, there exists `.ips` prop (ExpressJS would have it too, so I should have guessed), and its value would indeed replicate ExpressJS' `.ips` prop

But aside from that, ExpressJS would actually also set `.ip` prop to the following:
> When the `trust proxy` setting does not evaluate to `false`, the value of this property is derived from the left-most entry in the `X-Forwarded-For` header. This header can be set by the client or by the proxy.
> Source (v4.x API doc): https://expressjs.com/en/4x/api.html#req.ip

And I can also verify that exact behavior, as my previous codes that was based on ExpressJS used to rely on `req.ip` prop alone to simplify things
As in, I can rely on that single prop alone to always be filled with visitors' actual IP address, simply by toggling `trust proxy` setting according to whether I use or not use a proxy

**Reply by kartikk221** [reply]
I see, so essentially when `trust_proxy` is `true` then the `Response.ip` property automatically returns the first IP in the `X-Forwarded-For` chain?

**Reply by BobbyWibowo** [reply]
At least that's the case in ExpressJS, yes

**Reply by kartikk221** [reply]
I've pushed a commit that will return the first IP address from the "x-forwarded-for" header set of IPs if trust_proxy is enabled here: https://github.com/kartikk221/hyper-express/commit/23f0ad90dab5403e5d4bf2a89c44d3b01cddb79f If this properly matches the ExpressJS behavior then I'll go ahead and have it pushed in in the next update.

**Reply by BobbyWibowo** [reply]
Thanks, I will confirm tomorrow (I live on the other side of the planet)

**Reply by BobbyWibowo** [reply]
Hi there, just want to inform that 6.2.6 is behaving as expected now, thank you!
